include/networkit/randomization/EdgeSwitching.hpp

```diff
@@ -0,0 +1,138 @@
+/*
+ * EdgeSwitching.hpp
+ *
+ *  Created on: 19.10.2019
+ *      Author:  Manuel Penschuck <networkit@manuel.jetzt>
+ */
+// networkit-format
+
+#ifndef NETWORKIT_RANDOMIZATION_EDGE_SWITCHING_HPP_
+#define NETWORKIT_RANDOMIZATION_EDGE_SWITCHING_HPP_
+
+#include <random>
+#include <string>
+#include <utility>
+
+#include <networkit/Globals.hpp>
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/graph/Graph.hpp>
+
+namespace NetworKit {
+
+/**
+ * EdgeSwitching and EdgeSwitchingInPlace implement the same algorithm -- while the former copies
+ * the input graph provided and is more versatile, the in-place implementation only keeps a
+ * reference to the input graph.
+ *
+ * The Edge Switching Markov Chain ["The markov chain simulation method for generating connected
+ * power law random graphs", Mihail and Zegura] perturbates simple directed or undirected graphs
+ * while preserving their degrees. In each step, two edges are selected uniformly at random, and
+ * their endpoints exchanged. Swaps that introduce multi-edges or self-loops are rejected WITHOUT
+ * replacement -- this is necessary to allow uniform sampling [see "Switching edges to randomize
+ * networks: what goes wrong and how to fix it", Carstens and Horadam]. The number of successful
+ * swaps can be queried using getNumberOfAffectedEdges()/2.
+ *
+ * @note In general, simple edge switching does not yield a uniform distribution on simple DIRECTED
+ * graphs because the orientation of directed triangles cannot be changed. Using
+ * DegreePreservingShuffle as a preprocessing step overcomes this limitation. The preprocessing can
+ * also jump-start the perturbation process, yielding to potentially faster mixing.
+ *
+ * It's recommended to implement the preprocessing step in the calling code or to use copying
+ * EdgeSwitching implementation.
+ */
+class EdgeSwitchingInPlace : public Algorithm {
+public:
+    /**
+     * Constructs an EdgeSwitch algorithm that will change the input graph IN-PLACE.
+     * @warning For directed graphs preprocessing with DegreePreservingShuffle is necessary. Either
+     * do it manually, or use another constructor.
+     */
+    EdgeSwitchingInPlace(Graph &G) : graph(&G) {}
+
+    ~EdgeSwitchingInPlace() override = default;
+
+    /**
+     * Carry out 10*numberOfEdges() edge swaps.
+     * @note This function can be called several times. The graph's topology may not be changed
+     * externally between two invocations (this is relevant only for graphs not owned by the
+     * algorithm)
+     */
+    void run() override { run(10 * graph->numberOfEdges()); }
+
+    /**
+     * Carry out numberOfSwitches many edge swaps.
+     * @note This function can be called several times. The graph's topology may not be changed
+     * externally between two invocations (this is relevant only for graphs not owned by the
+     * algorithm)
+     */
+    void run(count numberOfSwitches);
```

Comment:
I would prefer an interface that accepts the number of switches in the constructor. Such an interface is more in line with the other algorithms that we have. We can expose a public `numberOfSwitches` member variable if this needs to be changed dynamically.

Reply:
As I do not have a strong opinion either way, I implemented it as suggest.